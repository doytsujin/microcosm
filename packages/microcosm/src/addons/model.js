/**
 * @fileoverview A general modelling class used by the Presenter's
 * getModel function.
 * @flow
 */

import { type Microcosm, Emitter, merge } from '../index'
import { type Job } from './default-update-strategy'

function isObservable(binding: *): boolean {
  return binding && typeof binding.subscribe === 'function'
}

function isCallable(binding: *): boolean {
  return binding && typeof binding.call === 'function'
}

type Answer = { [string]: * }

export default class Model extends Emitter {
  _scope: Object
  _repo: Microcosm
  _bindings: *
  _subscriptions: { [string]: * }
  _pendingUpdate: ?Job
  _publish: () => void
  _patch: Answer
  _watching: boolean

  value: Answer

  constructor(repo: Microcosm, scope: Object) {
    super()

    this._repo = repo
    this._scope = scope
    this._bindings = {}
    this._subscriptions = {}
    this._patch = {}
    this._publish = this._publish.bind(this)
    this._pendingUpdate = null
    this._watching = false

    this.value = {}
  }

  bind(bindings: Object) {
    let next = null

    for (var key in bindings) {
      var callback = bindings[key]

      if (isObservable(callback)) {
        // Presenters support observables. Updates to observables
        // assign to a single key/value pair
        this._track(key, callback)
      } else if (isCallable(callback)) {
        if (next === null) {
          next = {}
        }

        next[key] = callback
      } else if (this.value[key] !== callback) {
        this._patch[key] = callback
      }
    }

    this._rebind(next)

    this._computeNow()
  }

  /**
   * Dispose a model, removing all _subscriptions and unsubscribing
   * from the repo.
   */
  teardown() {
    for (var key in this._subscriptions) {
      this._subscriptions[key].unsubscribe()
    }

    this._empty()
    this._ignoreRepo()
  }

  /* Private ------------------------------------------------------ */

  _rebind(bindings: ?Object) {
    if (bindings) {
      this._bindings = bindings
      this._watchRepo()
    } else {
      this._bindings = {}
      this._ignoreRepo()
    }
  }

  _watchRepo() {
    if (!this._watching) {
      this._repo.on('change', this._computeEventually, this)
      this._watching = true
    }
  }

  _ignoreRepo() {
    if (this._watching) {
      this._repo.off('change', this._computeEventually, this)
      this._watching = false
    }
  }

  _enqueue() {
    if (!this._pendingUpdate) {
      this._pendingUpdate = this._repo.history.updater(this._publish)
    }
  }

  _set(key: string, value: *) {
    if (this.value[key] !== value) {
      this._patch[key] = value
      this._enqueue()
    }
  }

  _publish() {
    let next = merge(this.value, this._patch)

    if (next !== this.value) {
      this._emit('will-change', next, this._patch)
      this._commit(next)
      this._emit('change', next)
    }
  }

  _commit(value: Answer) {
    this.value = value
    this._empty()
  }

  _empty() {
    this._patch = {}

    if (this._pendingUpdate) {
      this._pendingUpdate.cancel()
      this._pendingUpdate = null
    }
  }

  /**
   * Track an observable. Sending updates to a given key.
   */
  _track(key: string, observable: *) {
    let last = this._subscriptions[key]
    let next = observable.subscribe(value => this._set(key, value))

    this._subscriptions[key] = next

    if (last) {
      last.unsubscribe()
    }
  }

  /**
   * Run through each callable binding, recomputing the model for
   * their associated keys.
   */
  _compute() {
    let dirty = false

    for (var key in this._bindings) {
      var callback = this._bindings[key]
      var value = callback.call(this._scope, this._repo.state, this._repo)

      if (this.value[key] !== value) {
        this._patch[key] = value
        dirty = true
      }
    }

    return dirty
  }

  _computeNow() {
    this._compute()
    this._publish()
  }

  _computeEventually() {
    if (this._compute()) {
      this._enqueue()
    }
  }
}
