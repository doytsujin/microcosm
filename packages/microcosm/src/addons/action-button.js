/**
 * @flow
 */

import React from 'react'
import Observable from 'zen-observable'
import $observable from 'symbol-observable'
import { merge } from '../index'

const identity = n => n

type Props = {
  action: *,
  onCancel: ?Callback,
  onClick: (event?: Event, Action?: *) => *,
  onDone: ?Callback,
  onError: ?Callback,
  onNext: ?Callback,
  onStart: ?Callback,
  onComplete: ?Callback,
  prepare: (value?: *, event?: Event) => *,
  send: ?Sender,
  tag: string | React$ElementType,
  value: *
}

type Context = {
  send: ?Sender
}

class ActionButton extends React.PureComponent<Props> {
  static defaultProps: Props
  static contextTypes: Context

  send: Sender
  click: (event: Event) => Action
  _queue: *[]

  constructor(props: Props, context: Context) {
    super(props, context)

    this.send = this.props.send || this.context.send
    this.click = this.click.bind(this)
    this._queue = []
  }

  componentWillUnmount() {
    this._queue.forEach(item => item.unsubscribe())
  }

  render() {
    const props = merge(this.props, { onClick: this.click })

    delete props.tag
    delete props.action
    delete props.value
    delete props.onStart
    delete props.onNext
    delete props.onComplete
    delete props.onError
    delete props.onNext
    delete props.send
    delete props.prepare

    if (this.props.tag === 'button' && props.type == null) {
      props.type = 'button'
    }

    return React.createElement(this.props.tag, props)
  }

  click(event: Event) {
    const { action, onClick, prepare, value } = this.props

    let params = prepare(value, event)
    let result = this.send(action, params)

    if (result && $observable in result) {
      this._queue.push(
        result.subscribe({
          start: this.props.onStart,
          next: this.props.onNext,
          complete: this.props.onComplete,
          error: this.props.onError
        })
      )
    }

    onClick(event, result)

    return result
  }
}

ActionButton.contextTypes = {
  send: () => {}
}

ActionButton.defaultProps = {
  action: null,
  onClick: identity,
  onStart: identity,
  onNext: identity,
  onComplete: identity,
  onError: identity,
  prepare: identity,
  send: null,
  tag: 'button',
  value: null
}

export default ActionButton
