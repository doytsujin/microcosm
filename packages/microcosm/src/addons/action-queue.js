/**
 * A commonized way to track a set of outstanding actions associated
 * with another entity, like an ActionForm or ActionButton.
 * @flow
 */

import { type Action } from '../index.js'

class ActionQueue {
  _queue: Action[]
  _scope: any

  constructor(scope: any) {
    this._queue = []
    this._scope = scope
  }

  size(): number {
    return this._queue.length
  }

  push(action: Action, callbacks: ?Object): void {
    if (callbacks) {
      action.subscribe(callbacks)
    }

    action.subscribe({})

    this._queue.push(action)
  }

  empty(): void {
    for (var i = 0, len = this._queue.length; i < len; i++) {
      this._queue[i]._removeScope(this._scope)
    }

    this._queue.length = 0
  }

  /* Private ------------------------------------------------------ */

  _remove(action: Action): void {
    this._queue.splice(this._queue.indexOf(action), 1)
  }

  _clean(): void {
    for (var i = 0, len = this._queue.length; i < len; i++) {
      var action = this._queue[i]

      if (action.complete) {
        this._remove(action)
      }
    }
  }
}

export default ActionQueue
