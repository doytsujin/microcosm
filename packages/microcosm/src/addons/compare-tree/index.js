/**
 * @flow
 * @private
 */

import Node from './node'
import Query from './query'
import { getKeyString, getKeyPaths, type KeyPath } from '../../key-path'
import { type Callback } from './emitter'

// The root key is an empty string. This can be a little
// counter-intuitive, so we keep track of them as a named constant.
const ROOT_PATH = ''

class CompareTree {
  snapshot: Object
  nodes: { [key: string]: Node }
  queries: { [key: string]: Query }

  constructor(initial: Object) {
    this.snapshot = initial
    this.nodes = {}
    this.queries = {}
  }

  /**
   * Create a subscription to a particular set of key paths.
   */
  on(subscription: *, callback: Callback, scope?: Object) {
    let keyPaths = getKeyPaths(subscription)
    let query = this.addQuery(keyPaths)

    for (var i = keyPaths.length - 1; i >= 0; i--) {
      this.addBranch(keyPaths[i], query)
    }

    query.on('change', callback, scope)

    return query
  }

  /**
   * Remove a subscription created by .on()
   */
  off(subscription: *, callback: Callback, scope?: Object) {
    let keyPaths = getKeyPaths(subscription)
    let id = Query.getId(keyPaths)

    let query = this.queries[id]

    if (query) {
      query.off('change', callback, scope)

      if (query.isAlone()) {
        this.prune(query)
      }
    }
  }

  /**
   * Compare a new snapshot to the last snapshot, triggering event
   * subscriptions along the way.
   */
  update(snapshot: Object) {
    let last = this.snapshot
    this.snapshot = snapshot

    let root = this.nodes[ROOT_PATH]

    if (root) {
      var queries = this.scan(this.nodes[ROOT_PATH], last, snapshot, [])

      for (var i = 0; i < queries.length; i++) {
        queries[i].trigger(snapshot)
      }
    }
  }

  /**
   * Add a node to the tree if it has not otherwise been added.
   */
  addNode(key: string, parent: ?Node) {
    let id = Node.getId(key, parent)

    if (this.nodes.hasOwnProperty(id) === false) {
      this.nodes[id] = new Node(id, key, parent)
    }

    return this.nodes[id]
  }

  /**
   * Add a query to the tree if it has not otherwise been
   * added. Queries are leaf nodes responsible for managing
   * subscriptions.
   */
  addQuery(dependencies: KeyPath[]): Query {
    let id = Query.getId(dependencies)

    if (typeof this.queries[id] === 'undefined') {
      this.queries[id] = new Query(id, dependencies)
    }

    return this.queries[id]
  }

  /**
   * Remove a query, then traverse that queries key paths to remove
   * unused parents.
   */
  prune(query: Query) {
    let ids = query.keyPaths.map(getKeyString)

    for (var i = 0, len = ids.length; i < len; i++) {
      let node = this.nodes[ids[i]]

      node.disconnect(query)

      do {
        if (node.isAlone()) {
          node.orphan()
          delete this.nodes[node.id]
        } else {
          break
        }

        node = node.parent
      } while (node)
    }

    delete this.queries[query.id]
  }

  /**
   * Build up a branch of nodes given a path of keys, appending a query
   * to the end.
   */
  addBranch(path: KeyPath, query: Query) {
    let last = this.addNode(ROOT_PATH, null)

    for (var i = 0, len = path.length; i < len; i++) {
      last = this.addNode(path[i], last)
    }

    last.connect(query)
  }

  /**
   * Traverse the tree of subscriptions, triggering queries along the way
   */
  scan(node: Node, last: *, next: *, queries: Query[]) {
    if (last !== next) {
      var edges = node.edges

      for (var i = 0, len = edges.length; i < len; i++) {
        var edge = edges[i]

        if (edge instanceof Query && queries.indexOf(edge) < 0) {
          queries.push(edge)
        } else if (edge instanceof Node) {
          var edgeLast = last == null ? last : last[edge.key]
          var edgeNext = next == null ? next : next[edge.key]

          this.scan(edge, edgeLast, edgeNext, queries)
        }
      }
    }

    return queries
  }
}

export default CompareTree
