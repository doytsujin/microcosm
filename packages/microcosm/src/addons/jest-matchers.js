/**
 * @flow weak
 */

import Microcosm, { get } from '../index'

declare var expect: any

expect.extend({
  toHaveStatus(action, status) {
    let operator = this.isNot ? 'not to' : 'to'
    let pass = action.status === status

    return {
      pass: pass,
      message: () => {
        return `Expected action ${operator} be '${status}'. Instead got ${
          action.status
        }.`
      }
    }
  },

  toHaveState(repo, key, value) {
    if (repo instanceof Microcosm === false) {
      throw new TypeError('toHaveState expects a Microcosm.')
    }

    let operator = this.isNot ? 'not to' : 'to'
    let pass = false
    let actual = get(repo.state, key)

    if (arguments.length > 2) {
      pass = JSON.stringify(actual) === JSON.stringify(value)
    } else {
      pass = actual !== undefined
    }

    // Display friendly key path
    let path = [].concat(key).join('.')

    return {
      pass: pass,
      message: () => {
        return (
          `Expected '${path}' in repo.state ${operator} be ${this.utils.printExpected(
            value
          )} ` + `but it is ${this.utils.printReceived(actual)}.`
        )
      }
    }
  }
})
