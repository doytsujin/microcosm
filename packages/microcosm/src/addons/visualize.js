import asciitree from 'asciitree'

export function asTree(history) {
  function dig(focus) {
    let children = history.children(focus).map(dig)

    return children.length ? [focus.tag, ...children] : focus.tag
  }

  console.log(asciitree(dig(history.root)))
}
