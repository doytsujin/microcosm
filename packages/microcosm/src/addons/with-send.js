/**
 * Connect a component to the presenter tree
 * @flow
 */

import { createElement } from 'react'
import { merge } from '../index'

function displayName(Component: Object) {
  return Component.displayName || Component.name || 'Component'
}

const CONTEXT_TYPES = {
  send: () => {}
}

export default function withSend(Component: *): * {
  function Sender(props: Object, context: Object) {
    let send = props.send || context.send

    console.assert(
      this.send,
      `${
        Sender.displayName
      } was not given \`send\` via context or props. Was this component mounted within a Presenter?`
    )

    return createElement(Component, merge({ send }, props))
  }

  Sender.displayName = 'withSend(' + displayName(Component) + ')'

  Sender.contextTypes = CONTEXT_TYPES

  Sender.WrappedComponent = Component

  return Sender
}
