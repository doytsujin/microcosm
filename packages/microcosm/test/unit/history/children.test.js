import Microcosm from 'microcosm'
import { asTree } from 'microcosm/addons/visualize'

describe('History node children', function() {
  it('can determine children', function() {
    let repo = new Microcosm({ debug: true })

    let a = repo.push('a')
    let b = repo.push('b')

    repo.history.checkout(a)

    let c = repo.push('c')

    expect(asTree(repo.history)).toMatchSnapshot()
  })

  it('does not lose children when checking out nodes on the left', function() {
    let repo = new Microcosm({ debug: true })

    repo.push('a')

    let b = repo.push('b')
    let c = repo.push('c')

    repo.history.checkout(b)

    let d = repo.push('d')

    expect(asTree(repo.history)).toMatchSnapshot()
  })

  it('does not lose children when checking out nodes on the right', function() {
    let repo = new Microcosm({ debug: true })

    repo.push('a')

    let b = repo.push('b')
    let c = repo.push('c')

    repo.history.checkout(b)

    let d = repo.push('d')

    repo.history.checkout(c)

    expect(asTree(repo.history)).toMatchSnapshot()
  })
})
