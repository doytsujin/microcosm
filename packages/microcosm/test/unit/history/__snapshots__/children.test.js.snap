// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`History node children can determine children 1`] = `
"    a
   / \\\\
  b   c"
`;

exports[`History node children does not lose children when checking out nodes on the left 1`] = `
"    a
    |
    b
   / \\\\
  c   d"
`;

exports[`History node children does not lose children when checking out nodes on the right 1`] = `
"    a
    |
    b
   / \\\\
  c   d"
`;
